### PR TITLE
New Feature: Halide Program IR Visualizer

### PR DESCRIPTION
**Pull request by darya-ver**
# New Feature: Halide Program IR Visualizer #

This PR is committing a new feature that generates an HTML page, with extension `.stmt.viz.html`, containing a visualization of Halide's IR. Here is an example of what the HTML page looks like: 

Screen Shot 2022-09-23 at 11 29 43 AM

You can also see the actual html page here: https://darya-ver.github.io/files/project_files/visualize_halide_example_blur3x3.html.

---
## File Changes ##

This PR is changing the following *existing files*:

* `Makefile`: added the new files into `SOURCE_FILES` and `HEADER_FILES`.

* `src/CodeGen_LLVM.h` // `src/CodeGen_LLVM.cpp`: added logic in `CodeGen_LLVM::visit(const ProducerConsumer *op)` and `CodeGen_LLVM::visit(const For *op)` to add unique identifiers to the markers in the generated LLVM code to more easily map the Halide IR code to the LLVM code.

* `src/Generator.cpp`: added logic for `-e stmt_viz` flag to generate the HTML page. Also added `assembly` flag if `stmt_viz` flag is present so that the assembly code is also generated.

* `src/Module.h` // `src/Module.cpp`: added new logic to call `Internal::print_to_viz()` if the `-e stmt_viz` flag is present.

This PR is adding the following *new files*:

* `src/StmtViz.h` // `src/StmtViz.cpp`: this is the main file that calls all other files. It contains the logic to generate the HTML page, and defines in the `print_to_viz()` function that is called from `src/Module.cpp`.

* `src/FindStmtCost.h` // `src/FindStmtCost.cpp`: added a new file that contains the logic to find the cost of a Halide statement. This is used to determine the color of the nodes in the visualization.

* `src/GetAssemblyInfoViz.h` // `src/GetAssemblyInfoViz.cpp`: added a new file that contains the logic to copy over the contents of the `.s` assembly file into a div in the HTML page, and finds the mapping of Halide IR to LLVM code.

* `src/GetStmtHierarchy.h` // `src/GetStmtHierarchy.cpp`: added a new file that contains the logic to generate the statement hierarchy of the nodes in the visualization.

* `src/IRVisualization.h` // `src/IRVisualization.cpp`: added a new file that contains the logic to generate the main visualization in the middle column.

---
## How to Navigate Webpage ##

Looking at the screenshot above, if you click the information button at the top right of the screen, you will see the following popup, which has information about what each column represents and what features exist within them. 

Screen Shot 2022-09-23 at 1 51 06 PM

---
## How to Generate HTML File ##

To use this new feature, your code needs to be a `Generator`. Once you compile the generator, run it with the `-e stmt_viz` flag. This will generate an HTML page in the directory that you've specified in the `-o` flag, with extension `.stmt.viz.html`.

For example, if you've compiled a generator called `my_generator` and you want to generate the HTML page in the `bin/` directory, you would run the following command:

```
./my_generator -e stmt_viz -o bin [rest of the flags]
```

where `[rest of flags]` are any another other flags that you need to add to correctly run the generator (like `-g`, `-f`, `target`, etc).

---
## TODO before merging into main ##

* Merge the `stmt_viz` and `stmt_html` flags into 1 so that we get rid of original IR visualization.

### CODE REVIEW

File: src/FindStmtCost.cpp
Comment (steven-johnson) [inline src/FindStmtCost.cpp:196]: Don't use dynamic_cast<> in Halide code -- we aren't always build with RTTI enabled, so this won't work -- and there is a *much* better way to do this with Halide IR nodes, namely, the `.as<>()` method (which is heavily optimized):

```
    const IfThenElse *if_then_else = op->as<IfThenElse>();  // nullptr if op is not of this type
```

File: src/FindStmtCost.cpp
Comment (steven-johnson) [inline src/FindStmtCost.cpp:235]: Do you really want to pass this vector by value here? Wouldn't it be preferable to pass as a const ref?

File: src/FindStmtCost.cpp
Comment (steven-johnson) [inline src/FindStmtCost.cpp:505]: Style nit, here and many other places: the first line of a function body should never be a blank line.

File: src/FindStmtCost.cpp
Comment (steven-johnson) [inline src/FindStmtCost.cpp:541]: The pattern of calling `set_costs` twice with the same args (except for the first) happens so many times that it seems worthy of its own little helper function.

File: src/FindStmtCost.cpp
Comment (steven-johnson) [inline src/FindStmtCost.cpp:581]: These should all be `internal_assert`

File: src/FindStmtCost.cpp
Comment (steven-johnson) [inline src/FindStmtCost.cpp:819]: Sure looks like a `switch` statement would be more appropriate here.

File: src/FindStmtCost.cpp
Comment (steven-johnson) [inline src/FindStmtCost.cpp:816]: nit: `ostringstream` might be slightly better since it's output-only

File: src/FindStmtCost.h
Comment (steven-johnson) [inline src/FindStmtCost.h:17]: Prefer to use either `constexpr int` or `enum` for numeric constants; use #define only when unavoidable

File: src/FindStmtCost.h
Comment (steven-johnson) [inline src/FindStmtCost.h:13]: `using namespace` in a .h file is strictly prohibited and should ~never be done.

File: src/FindStmtCost.h
Comment (steven-johnson) [inline src/FindStmtCost.h:11]: Be aware that Halide needs to be able to compile with exceptions enabled *and* disabled -- and I don't see what this is used for. Was it included accidentally?

File: src/Generator.cpp
Comment (steven-johnson) [inline src/Generator.cpp:854]: Hrm... not sure about this. I think it would be preferable to just fail with a message that specifying `stmt_viz` requires that `assembly` also be specified, if for no other reason than it makes it easier to integrate with Blaze/Bazel, which is very picky about undeclared files being generated. (Other Halide contributors may disagree here.)

File: src/GetAssemblyInfoViz.cpp
Comment (steven-johnson) [inline src/GetAssemblyInfoViz.cpp:1]: nit: no blank lines at the top of a source file

File: src/GetAssemblyInfoViz.cpp
Comment (steven-johnson) [inline src/GetAssemblyInfoViz.cpp:5]: In general, don't use `using namespace` in Halide source at all, at least inside the `src/` directory. This code should probably all be in the `Halide::Internal` namespace anyway, so the usings would be redundant.

File: src/GetAssemblyInfoViz.cpp
Comment (steven-johnson) [inline src/GetAssemblyInfoViz.cpp:71]: Why do we want a `./` prefix here -- isn't a plain name preferable?

File: src/GetAssemblyInfoViz.cpp
Comment (steven-johnson) [inline src/GetAssemblyInfoViz.cpp:89]: nit, here and elsewhere: don't have a blank line at the start of a braced block

File: src/GetAssemblyInfoViz.cpp
Comment (steven-johnson) [inline src/GetAssemblyInfoViz.cpp:133]: Do you need to use `regex` for this? Seems like the `replace_all()` function in Util.h would probably do the same thing a bit more cheaply

File: src/GetStmtHierarchy.cpp
Comment (steven-johnson) [inline src/GetStmtHierarchy.cpp:604]: This is fine as-is, but it might be cleaner to use a C++11 raw string literal instead.

File: src/GetStmtHierarchy.cpp
Comment (steven-johnson) [inline src/GetStmtHierarchy.cpp:629]: (1) you definitely want a raw string literal here
(2) declaring this as `const char *` or `const char[]` instead of `std::string` would avoid unnecessary memory allocations

File: src/GetStmtHierarchy.cpp
Comment (steven-johnson) [inline src/GetStmtHierarchy.cpp:645]: see previous comment

File: src/IRVisualization.cpp
Comment (steven-johnson) [inline src/IRVisualization.cpp:22]: `internal_error` will *never* return, so the `return StmtSize()` is unnecessary

**Comment by zvookin** [comment]
Wow. Very cool. Will try it out.

File: src/IRVisualization.cpp
Comment (steven-johnson) [inline src/IRVisualization.cpp:549]: nit: blank line between functions/methods

File: src/IRVisualization.cpp
Comment (steven-johnson) [inline src/IRVisualization.cpp:550]: nit, here and elsewhere: probably prefer `ostringstream` for a stream that is only written to

File: src/IRVisualization.cpp
Comment (steven-johnson) [inline src/IRVisualization.cpp:828]: ...what if var_name.size() < 2?

File: src/IRVisualization.cpp
Comment (steven-johnson) [inline src/IRVisualization.cpp:1243]: `ir_viz_js` would really fit Halide style better

File: src/IRVisualization.cpp
Comment (steven-johnson) [inline src/IRVisualization.cpp:1409]: this also should probably be `const char *` and a raw string literal

File: src/IRVisualization.cpp
Comment (steven-johnson) [inline src/IRVisualization.cpp:1443]: ditto

File: src/IRVisualization.h
Comment (steven-johnson) [inline src/IRVisualization.h:14]: Don't use #define for numeric constants -- `constexpr` or `enum`, please

File: src/StmtToViz.cpp
Comment (steven-johnson) [inline src/StmtToViz.cpp:32]: nit: no blank lines at the start of a struct/class declaration

File: src/StmtToViz.cpp
Comment (steven-johnson) [inline src/StmtToViz.cpp:114]: nit: consistently put one blank line between method bodies

File: src/StmtToViz.cpp
Comment (steven-johnson) [inline src/StmtToViz.cpp:357]: ... a comment about what this shape looks like would be nice :-)

File: src/StmtToViz.cpp
Comment (steven-johnson) [inline src/StmtToViz.cpp:464]: `std::string` is defined as being `char`, so this is doing an implicit conversion to `unsigned char`. Presumably no compiler complains?

File: src/StmtToViz.cpp
Comment (steven-johnson) [inline src/StmtToViz.cpp:486]: This is creating (and destroying) this string for *every character* you output. It should really be a `static const char [16]`.

File: src/IRVisualization.h
Comment (zvookin) [inline src/IRVisualization.h:14]: Is it possible to use ```constexpr``` for these?

File: src/IRVisualization.h
Comment (zvookin) [inline src/IRVisualization.h:28]: It would be great to do the doxygen style comments so the classes show up in the documentation.

File: src/IRVisualization.h
Comment (zvookin) [reply src/IRVisualization.h:14]: And if these do not need to be exported, possibly move them into the .cpp file where they are used.

**Review by steven-johnson — COMMENTED** [review]
At first glance, this looks amazing -- nice work! I did a quick skim, and have lots of style nits and a handful of real issues.  (Most important is probably that 'using namespace' should be avoided, *especially* in .h files, as it creates ambiguities in name resolution.)

There are a lot of long string literals containing HTML, JS, etc. -- I idly wonder if some/any/all of them would be better off as source-files-on-disk rather than embedded string literals, but don't have any strong feeling about that yet.

Also, we'll need to update the CMake build rules (those are actually the canonical rules) before landing.

**Comment by steven-johnson** [comment]
(please run the `run-clang-format.sh` script to apply a lot of our style requirements)

**Comment by steven-johnson** [comment]
Playing with the UI in Chrome, I can't seem to resize the column width by grabbing the green vertical dividers -- the cursor changes as though I can resize them there, but click-and-drag does nothing.

**Comment by zvookin** [comment]
After a quick glance here, this looks fairly well structured and isolated from the rest of the code. I'm probably not the best person to comment on the internals, but I'm not seeing anything that worries me.

I believe the new files will need to be added to CMakeLists.txt.

File: src/FindStmtCost.h
Comment (alexreinking) [reply src/FindStmtCost.h:17]: And if they can be moved to a single TU, then they should be, too.

File: src/StmtToViz.cpp
Comment (alexreinking) [inline src/StmtToViz.cpp:2385]: This should be printed only at some more-verbose-than-default `debug()` level.

File: src/FindStmtCost.cpp
Comment (slomp) [reply src/FindStmtCost.cpp:190]: It's been a while, but it looks like `as<T>` is no longer an interface of `IRNode`; I see one in `IRHandle`, though.

Comment (slomp) [reply src/FindStmtCost.cpp:190]: Is it just a matter of locally wrapping he `IRNode*` into an `IRHandle` just to have access to the `as<T>()` interface?

Comment (abadams) [reply src/FindStmtCost.cpp:190]: get_if_node_cost should take a const IfThenElse *, and the cast should happen in the caller as a C-style cast. If you wrap it in an IRHandle to use `as` it'll do a pointless refcount.

File: src/FindStmtCost.h
Comment (abadams) [inline src/FindStmtCost.h:15]: This is the style we use for preprocessor macros, and the scope is a bit broad. I suggest moving these inside StmtCost and changing them to NormalNodeDMC and similar, so that usage looks like StmtCost::NormalNodeDMC

File: src/FindStmtCost.cpp
Comment (slomp) [reply src/FindStmtCost.cpp:190]: How about something like `string GetReadWrite::print_node(const IRNode *node) const { ... }`?

**Comment by alexreinking** [comment]
Related issue: #5756

**Comment by steven-johnson** [comment]
This is requesting to merge from an external fork into a local branch, so I don't think we can get the buildbots to run on this correctly. Ideally we should convert this into an 'ordinary' PR to merge it into Halide `main` branch, but I'll leave that to the original author to deal with lest I mess something up. In the meantime, I've made a sub-branch of this (with no additional changes) in https://github.com/halide/Halide/pull/7062 to get the buildbots to run.

**Comment by slomp** [comment]
> This is requesting to merge from an external fork into a local branch, so I don't think we can get the buildbots to run on this correctly. Ideally we should convert this into an 'ordinary' PR to merge it into Halide `main` branch, but I'll leave that to the original author to deal with lest I mess something up. In the meantime, I've made a sub-branch of this (with no additional changes) in #7062 to get the buildbots to run.

Yup, the idea is to merge from a fork branch (Darya's own `StmtCost` feature branch) to the true Halide repo, but in a non-main branch (`darya-ver/ir-viz`). Once it's there and we address the first batch of reviews, we'll merge, and then create a PR from `darya-ver/ir-viz` to `main`, where we can get more comprehensive tests.

**Comment by slomp** [comment]
@steven-johnson And thanks for running clang-tidy too :-)

File: src/FindStmtCost.cpp
Comment (maaz139) [inline src/FindStmtCost.cpp:194]: Unclear why inclusive cost ignores ignores the else-case. Discussing the cost-model is perhaps tangential to this PR but we should add a comment explaining the intention here.

**Review by maaz139 — APPROVED** [review]
Addressed most of the comments from other reviews and re-ran the clang-tidy script.

I plan to do some more code revisions but I'll defer those for the next PR (into Halide main).

**Comment by abadams** [comment]
Please don't merge yet. I want to do a pass.

**Comment by abadams** [comment]
Two big things: 

There appear to be a bunch of unrelated wasm and python float16 changes in here.

The header files are missing file-level documentation and doxygen-style non-trivial class and method comments throughout.

**Comment by abadams** [comment]
My other big concern is that I don't think computing and showing approximate costs is useful enough to justify the maintenance cost of the big chunk of code that computes it. I can't think of a case where I would find it helpful.

**Comment by maaz139** [comment]
> There appear to be a bunch of unrelated wasm and python float16 changes in here.

I traced the changes in wasm and I believe they were pulled into Darya's fork from main. The `ir-viz` branch that this PR pushes the new code probably needs to be updated as well.

**Comment by abadams** [comment]
Oh I mistakenly thought this was a PR into main. Yeah we can merge then continue to work on it.

**Pull request merged**
